lib/record_store/provider.rb

```diff
@@ -128,6 +128,20 @@ def master_nameserver_for(zone_name)
 
         dns.getresource(zone_name, Resolv::DNS::Resource::IN::SOA).mname.to_s
       end
+
+      def retry_on_connection_errors(max_timeouts: 5, max_conn_resets: 5)
+        loop do
+          begin
+            return yield
+          rescue Net::OpenTimeout
+            raise if max_timeouts <= 0
+            max_timeouts -= 1
+          rescue Errno::ECONNRESET
+            raise if max_conn_resets <= 0
+            max_conn_resets -= 1
```

Comment:
Could it make sense to:
1/ Add a slight delay before retry through `sleep`, maybe an increasing delay between each retry (1, 2, 4, etc.) up to a maximum delay if we don't want to wait too long
2/ Add an output message along with the delay, so deployments pipelines that monitor outputs do not reach a timeout before the end of the deployment

Reply:
1) Yes.  Particularly after the connection reset.  I'll let it retry without a delay on OpenTimeout, though...as it had just timed out.  Burning through 5 timeouts could take some time, so the quicker, the better.
2) Will throw in a message.  This is a good catch.  I think I can model it off of what's done during rate limiting.